### PR TITLE
trace frames with np.ndarray

### LINKED ISSUE #109604: Different graphbreaks may make that NumPy code is traced or not
### 🐛 Describe the bug

In the code below, `graph_break` generates two graphs, while using a `print` skips the second graph arguing that `[DEBUG] skipping because no torch.* <resume in fn>`

```python
import torch
import numpy as np


@torch.compile(backend="eager")
def fn(x):
    x = x ** 2
    #torch._dynamo.comptime.graph_break()
    print("HI")
    return 2 * x


fn(np.arange(8))
```

I think this may be caused by the source attribution for ndarrays in dynamo.

cc @mruberry @rgommers @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @ysiraichi @Fidget-Spinner 

### Versions

master

### PR DESCRIPTION
**Pull request by ev-br**
Fixes #109604

Resubmit gh-109715 + several skips and small fixes to make tests pass.

The main fix here is by @ysiraichi : previously, dynamo did not resume tracing numpy ndarrays after a graph break.
While at it, fix several small issues Yukio's fix uncovers:

- graph break gracefully on numpy dtypes which do not map to torch.dtypes (uint16 etc)
- recognize array scalars in dynamo, treat them as 0D ndarrays
- make sure that iterating over torch.ndarray generates arrays not bare tensors

cc @mruberry @rgommers @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/test_tensorboard.py [hunk 1]
```diff
@@ -802,11 +804,16 @@ def test_scalar(self):
         self.assertIsInstance(res, np.ndarray) and self.assertEqual(res.shape, (1,))
         res = make_np(np.float16(1.00000087))
         self.assertIsInstance(res, np.ndarray) and self.assertEqual(res.shape, (1,))
-        res = make_np(np.float128(1.00008 + 9))
-        self.assertIsInstance(res, np.ndarray) and self.assertEqual(res.shape, (1,))
         res = make_np(np.int64(100000000000))
         self.assertIsInstance(res, np.ndarray) and self.assertEqual(res.shape, (1,))
 
+    @skipIfTorchDynamo
```
Comment (lezcano) [inline test/test_tensorboard.py:810]: We can probably fix this one rather easily.

File: test/test_sparse_csr.py [hunk 2]
```diff
@@ -1537,6 +1537,7 @@ def run_test_block_addmm_addmv(self,
     @parametrize("noncontiguous", [True, False])
     @skipCPUIfNoMklSparse
     @unittest.skipIf(not TEST_SCIPY, "SciPy not found")
+    @skipIfTorchDynamo("raises 'sparse matrix length is ambiguous; use getnnz()'")
```
Comment (lezcano) [inline test/test_sparse_csr.py:1540]: and I'd expect this one to be fixable by breaking properly on sparse tensors

Comment (ev-br) [reply test/test_sparse_csr.py:1540]: Probably. The exception is from scipy though. I'd prefer to kick it down to a follow-up.

File: test/test_tensorboard.py (hunk 1 above)
Comment (lezcano) [reply test/test_tensorboard.py:810]: No, the issue here is that we should break on `float128`s and we are not doing so atm.

Comment (ev-br) [reply test/test_tensorboard.py:810]: We are breaking on np.float128.
Running

```
(pytorch-dev) ev-br@qgpu3:~/repos/pytorch$ cat ma.py 
import torch
from torch._dynamo import config
import torch._dynamo.testing
import numpy as np


def fn(x):
    return np.float128(x)

print(fn(1e100))


cnts = torch._dynamo.testing.CompileCounter()
opt_fn = torch._dynamo.optimize(cnts)(fn)

print(opt_fn(1e100))
print(cnts.frame_count)
```

produces

```
(pytorch-dev) ev-br@qgpu3:~/repos/pytorch$ TORCH_LOGS="graph_breaks" python ma.py 
1.0000000000000000159e+100
[2023-10-04 14:09:17,525] [0/0] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] Graph break: Can't find numpy function <class 'numpy.float128'> in torch._numpy.  Please file an issue to request support for this function. from user code at:
[2023-10-04 14:09:17,525] [0/0] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "ma.py", line 29, in fn
[2023-10-04 14:09:17,525] [0/0] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     return np.float128(x)
[2023-10-04 14:09:17,525] [0/0] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] 
1.0000000000000000159e+100
0
```

Comment (ev-br) [reply test/test_tensorboard.py]: Okay, this is a bit trickier. The example above is correct, but. If `np.float128(42)` is used as input to a function, it was not recognized as a numpy variable and eventually there was a hard failure at https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/builder.py#L1230.

One solution, which this PR does, is to make `is_numpy` function aware of array scalars and wrap them on an equal footing as ndarrays. This is what `torch._numpy` assumes anyway, so it should it be safe I expect.

File: test/test_tensorboard.py [hunk 3]
```diff
@@ -802,11 +803,15 @@ def test_scalar(self):
...
-        res = make_np(np.float128(1.00008 + 9))
-        self.assertIsInstance(res, np.ndarray) and self.assertEqual(res.shape, (1,))
         res = make_np(np.int64(100000000000))
         self.assertIsInstance(res, np.ndarray) and self.assertEqual(res.shape, (1,))
 
+    @unittest.skipIf(IS_WINDOWS, "Skipping on windows, see https://github.com/pytorch/pytorch/pull/109349 ")
+    @unittest.skipIf(IS_MACOS, "Skipping on mac, see https://github.com/pytorch/pytorch/pull/109349 ")
+    def test_scalar_f128(self):
+        res = make_np(np.float128(1.00008 + 9))
+        self.assertIsInstance(res, np.ndarray) and self.assertEqual(res.shape, (1,))
```
Comment (ev-br) [inline test/test_tensorboard.py:813]: TODO: If this test passes on CI, fold it back to `test_scalar`.

File: torch/_dynamo/variables/builder.py [hunk 4]
```diff
@@ -1114,7 +1114,15 @@ def wrap_numpy_ndarray(self, value):
         assert isinstance(value, np.ndarray)
 
         source = NumpyTensorSource(self.get_source())
-        tensor_value = torch.as_tensor(value)
+
+        from torch._numpy import _util
+
+        try:
+            tensor_value = _util._coerce_to_tensor(value)
```
Comment (ev-br) [inline torch/_dynamo/variables/builder.py:1121]: This sets a precedent of reaching into `torch._numpy` from the dynamo wrapper layer. In this particular case it's not a big deal to roll it back; in general it seems reasonable to try reducing the duplication between these two layers (dynamo wrapping into faketensors/numpyvariables and torch._numpy.ndarray wrapping tensors).

File: test/test_tensorboard.py (hunk 1 above)
Comment (lezcano) [reply test/test_tensorboard.py]: Wrapping NumPy scalars into arrays is good for now. Something we should do in the future is to implement proper scalar semantics on NumPy scalars, even if internally we represent them as a tensor.

One day...

File: torch/_dynamo/variables/builder.py (hunk 4 above)
Comment (lezcano) [reply torch/_dynamo/variables/builder.py:1121]: Our `torch._numpy` is a tool already used by dynamo, so this is completely fine, and if anything, the right thing to do.

File: test/dynamo/test_misc.py [hunk 5]
```diff
@@ -1320,9 +1320,10 @@ def fn(x, a, b):
             return x + (a % b)
 
         args = [torch.randn(10), 4096, np.int64(8)]
+#        args = [torch.randn(10), 4096, np.array(8, dtype='int64')]
```
Comment (lezcano) [inline test/dynamo/test_misc.py:1323]: uncomment?

File: test/test_tensorboard.py (hunk 3 above)
Comment (lezcano) [reply test/test_tensorboard.py:813]: passes ci so move

File: torch/_dynamo/variables/builder.py [hunk 6]
```diff
@@ -521,6 +521,9 @@ def index_source(key):
                 source=self.source,
                 guards=make_guards(GuardBuilder.ID_MATCH),
             )
+        elif isinstance(value, np.generic):
+            # numpy array scalars: convert to 0D arrays
+            return self.wrap_numpy_ndarray(np.asarray(value))
```
Comment (lezcano) [inline torch/_dynamo/variables/builder.py:526]: Actually, I think a better fix is to put this here:
https://github.com/pytorch/pytorch/blob/8a8e9a6c5f5b6103e7e877b8c0b27aca7b8323db/torch/_dynamo/variables/builder.py#L306
and then teach `wrap_numpy_ndarray` to handle np.generics.

Comment (lezcano) [reply torch/_dynamo/variables/builder.py:526]: better is not the word. it's just that it'd be together with its friend np.ndarray

Comment (ev-br) [reply torch/_dynamo/variables/builder.py:526]: as is it's next to `elif is_numpy()` (next line)

File: test/test_tensorboard.py (hunk 1 above)
Comment (ev-br) [reply test/test_tensorboard.py]: Discussed this offline with Mario, and the current state of the PR does not skip this test anymore (was a bit of a rabbit hole but hey).

File: test/dynamo/test_unspec.py [hunk 7]
```diff
@@ -59,6 +59,7 @@ def fn(x, y):
         self.assertEqual(cnts.frame_count, 1)
         self.assertEqual(cnts.op_count, 2)
 
+    @unittest.skipIf(True, "array scalars decay to 0D arrays")
```
Comment (lezcano) [inline test/dynamo/test_unspec.py:62]: Nit. Let's xfail these so that we can remove this once we fix this.

**Review by lezcano — COMMENTED** [review]
The approach LGTM. There is a relevant error in CI tho.

File: torch/_dynamo/variables/tensor.py [hunk 8]
```diff
@@ -1019,6 +1019,36 @@ def call_method(
     def python_type(self):
         return np.ndarray
 
+    def unpack_var_sequence(self, tx, idxes=None):
+        # XXX: a clone of TensorVariable method, w/ wrap_fx_proxy -> wrap_fx_proxy_cls
```
Comment (lezcano) [inline torch/_dynamo/variables/tensor.py:1023]: not cool. Modify the TensorVariable method to call wrap_fx_proxy_cls with `type(self)`.

File: torch/_dynamo/variables/tensor.py [hunk 9]
```diff
@@ -1020,6 +1025,9 @@ def python_type(self):
         return np.ndarray
 
 
+s
```
[the remark this answers is not in the document]
Comment (ev-br) [reply torch/_dynamo/variables/tensor.py]: Yes, network latency makes ctrl-S infuriating.

File: torch/_dynamo/variables/tensor.py (hunk 8 above)
Comment (ev-br) [reply torch/_dynamo/variables/tensor.py]: Srlsly? lol, OK, fine. done.

**Comment by lezcano** [comment]
The dynamo error seems real

**Comment by lezcano** [comment]
I'll have a look at the graphbreaks

File: test/test_torch.py [hunk 10]
```diff
@@ -2909,7 +2909,7 @@ def test_gradient_extreme_cases(self, device, dtype):
         large_size = 100000
         t = make_tensor((large_size,), dtype=dtype, device=device)
         t_np = t.cpu().numpy()
-        coordinates_np = list(np.random.randn(large_size))
+        coordinates_np = np.random.randn(large_size)
```
Comment (lezcano) [inline test/test_torch.py:2912]: what's the issue with casting it to a list?

Comment (ev-br) [reply test/test_torch.py:2912]: Original numpy : `coordinates_np` becomes a list of array scalars; the next line, `tensor(coordinates_np)`,  converts it into a tensor.
Under dynamo: `coordinates_np` becomes a list of 0D `torch.ndarray`s; the next line, `tensor(coordinates_np)` fails with "len() of an unsized object".

Comment (lezcano) [reply test/test_torch.py:2912]: rrright. Thank you!

**Review by lezcano — APPROVED** [review]
The increased number of graphbreaks is expected given that now trace that forloop and we need to break in it. Could you update the relevant files and merge this one

**Comment by atalman** [comment]
will be including it in 2.2 release

**Pull request closed without merging**